thriftrw/spec/struct.pyx

```diff
@@ -370,41 +373,63 @@ def struct_init(cls_name, field_names, field_defaults, base_cls, validate):
                 )
             )
 
-        # Keeps track of fields yet to be assigned.
-        unassigned = set(field_names)
+        cdef object value
+        cdef FieldSpec field_spec
+        cdef int i = -1
+        for name in field_names:
+            i += 1
+            # Check positionals
+            if i < num_args:
+                value = args[i]
 
-        # Assign positional arguments.
-        for name, value in zip(field_names, args):
-            if value is not None:
-                setattr(self, name, value)
-                unassigned.remove(name)
+                # If found in positionals, it can't be in kwargs
+                if name in kwargs:
+                    raise TypeError(
+                        '%s() got multiple values for argument "%s"'
+                        % (cls_name, name)
+                    )
+            else:
+                value = kwargs.pop(name, None)
 
-        # Assign named arguments.
-        for name in field_names:
-            # We go over all field names instead of just the remaining ones so
-            # that we can detect if any of the positional arguments had
-            # another value specified in kwargs. That's user error and we
-            # should blow up in that case.
-            if name not in kwargs:
-                continue
-            if name not in unassigned:
-                raise TypeError(
-                    '%s() got multiple values for argument "%s"'
-                    % (cls_name, name)
-                )
-            value = kwargs.pop(name)
-            if value is not None:
-                setattr(self, name, value)
-                unassigned.remove(name)
-
-        # Add defaults
-        if field_defaults:
-            default_values = zip(field_names[-num_defaults:], field_defaults)
-            for name, value in default_values:
-                if name in unassigned:
-                    setattr(self, name, copy.deepcopy(value))
-                    unassigned.remove(name)
+            # Didn't find anything. Either supply a default or error.
+            if value is None:
+                if i >= begin_defaults_i:
+                    value = copy.deepcopy(field_defaults[i - begin_defaults_i])
+                else:
+                    raise ValueError(
+                        'Field %r requires a non-None value.' % name
```

Comment:
This is sort of unnecessary if you're already checking
`field_spec.required` below.

In fact, you can also apply the default below when
inspecting the `field_spec`, making the `field_defaults`
argument unnecessary. You'll have to pre-calculate
`num_required` outside the `__init__`.